### PR TITLE
Treat error response payloads from Slack as errors

### PR DESCRIPTION
**Pull request by knyar**
As described in the "More error types" section below, Slack API can return errors with a 200 response code:
https://slack.dev/node-slack-sdk/web-api#handle-errors

This change adds parsing of API response to extract error messages.

Signed-off-by: Anton Tolchanov <anton@tailscale.com>

### CODE REVIEW

**Comment by knyar** [comment]
@simonpasquier are you the best person to review this?

**Comment by knyar** [comment]
Just rebased to make this merge-able.

@gotjosh, @simonpasquier – any thoughts on this?

File: notify/slack/slack.go [hunk 1]
```diff
@@ -206,12 +213,36 @@ func (n *Notifier) Notify(ctx context.Context, as ...*types.Alert) (bool, error)
 	if err != nil {
 		return true, notify.RedactURL(err)
 	}
-	defer notify.Drain(resp)
+	defer resp.Body.Close()
 
 	// Only 5xx response codes are recoverable and 2xx codes are successful.
```
Comment (simonpasquier) [inline notify/slack/slack.go:218]: the comment would need an update.

**Review by simonpasquier — COMMENTED** [review]
is there any way that I can reproduce the issue locally?

File: notify/slack/slack.go (hunk 1 above)
Comment (knyar) [reply notify/slack/slack.go]: Removed this one and added a comment clarifying that sometimes errors come with a 200 response code.

**Comment by knyar** [comment]
> is there any way that I can reproduce the issue locally?

Sure:

```
$ curl -w "\ncode: %{http_code}\n" -X POST 'https://slack.com/api/chat.postMessage'
{"ok":false,"error":"not_authed"}
code: 200
```

**Comment by simonpasquier** [comment]
@gotjosh the grafana/alerting code is AGPL, right?
I'm not a lawyer but how are we sure that it won't create an issue?

**Comment by gotjosh** [comment]
This situation happened a couple of months ago and I consulted internally what were the steps necessary to allow that.

The result is https://github.com/prometheus/alertmanager/issues/2845#issuecomment-1303341819 as an example.

I'll use similar wording as I approve the PR.

**Comment by knyar** [comment]
> This is something that the Grafana equivalent does much better, can we instead borrow the idea from what we do there? 

Sure! Given alertmanager's use of common `Retrier` and `ErrorWithReason` it's not particularly straightforward here to re-use parts of that code in a meaningful way, but I have extended this PR to handle plaintext (non-JSON) responses as well. I have not added parsing of string error ids (like `channel_is_archived`) because they seem to be expressive enough.

Also expanded the test to cover:
- 1xx HTTP responses;
- Plaintext 2xx successful and error responses;
- JSON parsing errors;
- Classification of errors as retryable;
- Text of error messages.

I believe functionally this now should align with what Grafana are doing. PTAL?

File: notify/slack/slack.go [hunk 2]
```diff
@@ -210,16 +210,66 @@ func (n *Notifier) Notify(ctx context.Context, as ...*types.Alert) (bool, error)
 	if err != nil {
 		return true, notify.RedactURL(err)
 	}
-	defer notify.Drain(resp)
+	defer resp.Body.Close()
```
Comment (grobinson-grafana) [inline notify/slack/slack.go:213]: I'm not 100% familiar with `notify.Drain`, but why would we stop using it here? The call is deferred, so we can still read `resp.Body` from the function call on Line 226.

File: notify/slack/slack.go [hunk 3]
```diff
@@ -210,16 +210,66 @@ func (n *Notifier) Notify(ctx context.Context, as ...*types.Alert) (bool, error)
...
+	return checkTextResponseError(body)
+}
+
+// checkTextResponseError classifies plaintext responses from Slack.
+// A plaintext (non-JSON) response is successful if it's a string "ok".
+// This is typically a response for an Incoming Webhook
+// (https://api.slack.com/messaging/webhooks#handling_errors)
+func checkTextResponseError(body []byte) (bool, error) {
+	if !bytes.Equal(body, []byte("ok")) {
+		return false, fmt.Errorf("received an error response from Slack: %s", string(body))
```
Comment (grobinson-grafana) [inline notify/slack/slack.go:254]: This should be `true` shouldn't it? Don't we want to retry failed webhooks?

```suggestion
		return true, fmt.Errorf("received an error response from Slack: %s", string(body))
```

File: notify/slack/slack.go [hunk 4]
```diff
@@ -210,16 +210,66 @@ func (n *Notifier) Notify(ctx context.Context, as ...*types.Alert) (bool, error)
...
+func checkJSONResponseError(body []byte) (bool, error) {
+	// response is for parsing out errors from the JSON response.
+	type response struct {
+		OK    bool   `json:"ok"`
+		Error string `json:"error"`
+	}
+
+	var data response
+	if err := json.Unmarshal(body, &data); err != nil {
+		return false, errors.Wrapf(err, "could not unmarshal JSON response %q", string(body))
```
Comment (grobinson-grafana) [inline notify/slack/slack.go:269]: Same question here, earlier we retried if the body could not be read:

```go
body, err := io.ReadAll(resp.Body)
if err != nil {
	return true, errors.Wrap(err, "could not read response body")
}
```

Shouldn't we retry if the JSON cannot be unmarshalled too?

```suggestion
		return true, errors.Wrapf(err, "could not unmarshal JSON response %q", string(body))
```

Comment (grobinson-grafana) [reply notify/slack/slack.go:269]: Oh I see, in Grafana we don't retry these errors.

File: notify/slack/slack.go (hunk 3 above)
Comment (grobinson-grafana) [reply notify/slack/slack.go:254]: Oh I see, in Grafana we don't retry these errors.

File: notify/slack/slack.go (hunk 4 above)
Comment (gotjosh) [reply notify/slack/slack.go:268]: On the retry side, I think we should keep Alertmanager's behaviour intact (for now) - we can discuss as part of a separate PR if we should retry or not.

File: notify/slack/slack.go (hunk 2 above)
Comment (grobinson-grafana) [reply notify/slack/slack.go]: Thank you!

File: notify/slack/slack.go (hunk 3 above)
Comment (grobinson-grafana) [reply notify/slack/slack.go:253]: Yes I agree! Sorry for the original confusing comment!

**Comment by grobinson-grafana** [comment]
@gotjosh @simonpasquier something I missed is that these changes update Alertmanager to understand error responses from both Slack's [`chat.postMessage`](https://api.slack.com/methods/chat.postMessage) API and [incoming webhooks](https://api.slack.com/messaging/webhooks). These are two different APIs with different authentication schemes and different features.

Alertmanager supports both, but the fact that it does so seems coincidental rather than intentional (for example `slack_configs` contains `http_config` so `Bearer` token can be set for `chat.postMessage`). This is also the case in Grafana, and I have suggested to the team in the past to separate Slack into two receivers: one for `chat.postMessage` and another for incoming webhooks.

I think we should accept this change because I suspect Alertmanager users are using this integration for both APIs. However, I want to highlight that adding error handling for `chat.postMessage` further encourages the use of this integration for either API, despite the documentation stating that this integration is [**just for incoming webhooks**](https://prometheus.io/docs/alerting/latest/configuration/#slack_config).

File: notify/slack/slack.go (hunk 4 above)
Comment (grobinson-grafana) [reply notify/slack/slack.go:268]: I took a quick look and we don't have a precedence in Alertmanager for these specific kinds of errors. I'm inclined to return `true` if parsing the response JSON fails, as that's what we also do if reading the response buffer fails.

**Comment by grobinson-grafana** [comment]
LGTM once https://github.com/prometheus/alertmanager/pull/3121#discussion_r1286974714 has been fixed! Thanks for your contribution! ❤️

**Comment by gotjosh** [comment]
> further encourages the use of this integration for either API, despite the documentation stating that this integration is [just for incoming webhooks](https://prometheus.io/docs/alerting/latest/configuration/#slack_config).

@grobinson-grafana would you be OK With updating the documentation for that?

**Comment by grobinson-grafana** [comment]
Sure! I just wanted to confirm that's what we want 🙂

**Comment by grobinson-grafana** [comment]
I just tested a couple of different scenarios.

1. `text/plain` response with "oops"

```
ts=2023-08-10T09:27:48.060Z caller=dispatch.go:352 level=error component=dispatcher msg="Notify for alerts failed" num_alerts=1 err="slack/slack[0]: notify retry canceled due to unrecoverable error after 1 attempts: channel \"#test-alerts\": received an error response from Slack: oops\n"
```

2. `application/json` response with invalid JSON "oops"

```
ts=2023-08-10T09:28:29.707Z caller=notify.go:745 level=warn component=dispatcher receiver=slack integration=slack[0] aggrGroup={}:{} msg="Notify attempt failed, will retry later" attempts=7 err="channel \"#test-alerts\": could not unmarshal JSON response \"oops\\n\": invalid character 'o' looking for beginning of value"
```

3. `application/json` response with valid JSON

```
ts=2023-08-10T09:29:16.225Z caller=dispatch.go:352 level=error component=dispatcher msg="Notify for alerts failed" num_alerts=1 err="slack/slack[0]: notify retry canceled due to unrecoverable error after 9 attempts: channel \"#test-alerts\": error response from Slack: oops"
```

**Comment by grobinson-grafana** [comment]
PR to update the docs https://github.com/prometheus/alertmanager/pull/3455

**Review by gotjosh — APPROVED** [review]
LGTM

Thank you very much for your contribution.

**Pull request merged**